### PR TITLE
net: Introduce interfaces stability test

### PR DESCRIPTION
**Pull request by azhivovk**
Mulitple users have encountered events in which
vmi interfaces or ipAddress field disappeared.

VMI interface status is unstable during boot - interfaces may
temporarily disappear from the status until all are stable. 
Asserting before interfaces stabilize risks a false negative - during setup
we wait for all VM interfaces to stabilize before verifying interfaces stability.

During tests, using watch kuberenetes API, which
is used by wrapper's watcher, we monitor for such
events and verify all interfaces are stable.

## Summary by CodeRabbit

* **Tests**
  * Added optional IPv6 DHCP support in network interface test data.
  * Added IP persistence validation for Linux-bridge network setups.
  * Added tests ensuring VM network interfaces retain IP addresses after migration.

### CODE REVIEW

File: tests/network/vmi/libippersistence.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+from libs.vm.vm import BaseVirtualMachine
+from tests.network.libs import cloudinit
+from tests.network.libs.ip import random_ipv4_address, random_ipv6_address
+
+LOGGER = logging.getLogger(__name__)
+TIMEOUT_SECONDS = 300
+
+
+def vm_cloud_init_data(
```
Comment (EdDev) [inline tests/network/vmi/libippersistence.py]: I suggest you do not mix the primary and secondary interfaces configuration.
The primary better be located in a lib, as it is usually static (same IP, unless the subnet is changed explicitly).

The secondary can be placed here or in a general lib that will serves any potential tests like these. But starting locally is fine.

File: tests/network/vmi/libippersistence.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import time
+
+from kubernetes import watch
+
+from libs.vm.vm import BaseVirtualMachine
+from tests.network.libs import cloudinit
+from tests.network.libs.ip import random_ipv4_address, random_ipv6_address
+
+LOGGER = logging.getLogger(__name__)
+TIMEOUT_SECONDS = 300
```
Comment (EdDev) [inline tests/network/vmi/libippersistence.py]: Timeout of what?

File: tests/network/vmi/libippersistence.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            interfaces = vmi_obj.status.interfaces
+            assert interfaces, f"VMI {vmi_name} has no interfaces{context_msg}"
+            assert len(interfaces) == 2, (
+                f"Expected 2 interfaces, found {len(interfaces)} on VMI {vmi_name}{context_msg}"
+            )
+
+            for interface in interfaces:
+                assert interface.ipAddress, (
+                    f"ipAddress field missing on interface {interface.name} of VMI {vmi_name}{context_msg}"
+                )
```
[the remark this answers is not in the document]
Comment (EdDev) [reply tests/network/vmi/libippersistence.py]: I agree.

File: tests/network/vmi/libippersistence.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    ipv6_supported_cluster: bool,
+) -> cloudinit.NetworkData:
+
+    ethernets = {}
+
+    # Configure primary interface (eth0/masquerade)
+    if ipv6_supported_cluster:
+        ethernets["eth0"] = cloudinit.EthernetDevice(
+            addresses=["fd10:0:2::2/120"],
+            gateway6="fd10:0:2::1",
```
Comment (yossisegev) [inline tests/network/vmi/libippersistence.py]: These 2 values will have to be stored in some global argument. as they are already used in at least one other test.
But this doesn't have to happen in this PR.

File: tests/network/vmi/libippersistence.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    vmi_resource = vm.vmi.client.resources.get(api_version="kubevirt.io/v1", kind="VirtualMachineInstance")
+
+    try:
+        for event in vmi_resource.watch(
+            namespace=namespace,
+            name=vmi_name,
+            timeout=TIMEOUT_SECONDS,
+            watcher=watcher,
+        ):
+            iteration += 1
```
[the remark this answers is not in the document]
Comment (yossisegev) [reply tests/network/vmi/libippersistence.py]: @EdDev I think it's beneficial for logging purposes.
As it is not explicitly clear what is the interval of `watch()`, we can't say how many times the object was sampled and found to be valid. Having the value of `iteration` in the success log message fills this gap IMO.

File: tests/network/vmi/conftest.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    namespace: Namespace,
+    unprivileged_client: DynamicClient,
+    bridge_network: NetworkAttachmentDefinition,
+    ipv4_supported_cluster: bool,
+    ipv6_supported_cluster: bool,
+) -> Generator[BaseVirtualMachine]:
+    spec = base_vmspec()
+    spec.template.spec.domain.devices.interfaces = [  # type: ignore
+        Interface(name="default", masquerade={}),
+        Interface(name="linux-bridge", bridge={}),
```
Comment (Anatw) [inline tests/network/vmi/conftest.py]: from utilities.constants import LINUX_BRIDGE

**Review by Anatw — CHANGES_REQUESTED** [review]

File: tests/network/vmi/conftest.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+from libs.vm.factory import base_vmspec, fedora_vm
+from libs.vm.spec import Interface, Multus, Network
+from libs.vm.vm import BaseVirtualMachine
+from tests.network.vmi.libippersistence import vm_cloud_init_data
+from utilities.constants import LINUX_BRIDGE, WORKER_NODE_LABEL_KEY
+from utilities.network import network_device, network_nad
+
+
+@pytest.fixture(scope="module")
+def ip_persistence_bridge_device_name(index_number: Generator[int]) -> str:
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/vmi/conftest.py]: Done

**Comment by azhivovk** [comment]
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/05c2d7e6b74ea4a32e12b980ca44dafe5a0a15f3..2e284c839a10ab4e8714791604f6f176c9895636): Small fix

File: tests/network/non_functional/ip_persistence/conftest.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+
+@pytest.fixture(scope="module")
+def linux_bridge_vm_for_ip_persist(
+    ipv4_supported_cluster: bool,
+    ipv6_supported_cluster: bool,
+    unprivileged_client: DynamicClient,
+    namespace: Namespace,
+    bridge_network: NetworkAttachmentDefinition,
+) -> Generator[BaseVirtualMachine]:
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/non_functional/ip_persistence/conftest.py]: We already discussed this, redundant

File: tests/network/vmi/conftest.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        nad_name="test-bridge-network",
+        interface_name=ip_persistence_bridge_device_name,
+        namespace=namespace,
+        add_resource_name=False,
+    ) as nad:
+        yield nad
+
+
+@pytest.fixture(scope="module")
+def vm_single_nic_with_pod(
```
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/vmi/conftest.py]: >I’d prefer to follow the usual pattern we have and name the fixture after the scenario it supports, e.g. vm_for_ip_flapping or ip_persistence_vm

I don't see we have this pattern. Based on this logic, we should also fix all related fixture names and use in the naming the goal of the fixture, not a state ( for example `nncp_for_vm_ip_flapping`). I don't think this is a right way.

It also contradicts the e2e action, from user perspective what we do: "Take a VM with secondary iface and do some action (can be any)", but we don't do: "Take an IP flapping VM and do some action", it sounds senseless. The scenario itself is showing that as a setup we must have a "a VM with secondary iface".

File: tests/network/non_functional/ip_persistence/libippersistence.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            )
+            break
+
+        if event["type"] != "MODIFIED":
+            continue
+
+        interfaces = event["object"].status.interfaces
+        LOGGER.info(f"Event: VMI {vmi.name} status updated{context_msg}, interfaces: {interfaces}")
+
+    yield interfaces
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/non_functional/ip_persistence/libippersistence.py]: We always expect 2 interfaces here
If this test will require changes, the user will change it accordingly

File: tests/network/vmi/conftest.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/vmi/conftest.py]: Are we fine with the current name or should I rename it? `linux_bridge_vm_for_ip_persist`

File: tests/network/non_functional/ip_persistence/test_ip_persistence.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        assert interfaces, "VMI has no interfaces"
+        assert len(interfaces) == 2, f"Expected 2 interfaces, got {len(interfaces)}"
+        for iface in interfaces:
+            assert iface.ipAddress, f"ipAddress missing on interface {iface.name}"
+
+
+@pytest.mark.polarion("CNV-14340")
+def test_ip_address_persistence_after_migration(linux_bridge_vm_for_ip_persist):
+    migrate_vm_and_verify(vm=linux_bridge_vm_for_ip_persist)
+    LOGGER.info(
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/non_functional/ip_persistence/test_ip_persistence.py]: [I commented that I'll drop it](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930#discussion_r2846621755)

File: tests/network/non_functional/ip_persistence/libippersistence.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+) -> cloudinit.EthernetDevice:
+    addresses = []
+    if ipv4_supported_cluster:
+        addresses.append(f"{random_ipv4_address(net_seed=0, host_address=1)}/24")
+    if ipv6_supported_cluster:
+        addresses.append(f"{random_ipv6_address(net_seed=0, host_address=1)}/64")
+    return cloudinit.EthernetDevice(addresses=addresses)
+
+
+def ip_persist_vm(
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/non_functional/ip_persistence/libippersistence.py]: Renamed and will add to a follow-up

**Review by Anatw — APPROVED** [review]

File: tests/network/non_functional/ip_persistence/conftest.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    LINUX_BRIDGE_IFACE_NAME,
+    ip_persist_vm,
+)
+from utilities.constants import LINUX_BRIDGE, WORKER_NODE_LABEL_KEY
+
+IP_PERSISTENCE_BRIDGE_DEVICE_NAME = "br1-test"
+
+
+@pytest.fixture(scope="module")
+def bridge_devices(
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/non_functional/ip_persistence/conftest.py]: Done

File: tests/network/non_functional/ip_persistence/conftest.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+@pytest.fixture(scope="module")
+def bridge_nad(
+    admin_client: DynamicClient,
+    namespace: Namespace,
+    bridge_nncp: Generator[libnncp.NodeNetworkConfigurationPolicy],
+) -> Generator[NetworkAttachmentDefinition]:
+    config = NetConfig(
+        name="test-bridge-network",
+        plugins=[CNIPluginBridgeConfig(bridge=IP_PERSISTENCE_BRIDGE_DEVICE_NAME)],
```
Comment (EdDev) [inline tests/network/non_functional/ip_persistence/conftest.py]: I won't insist, but you could have fetched it from the `bridge_nncp` instead of defining a constant.

File: tests/network/non_functional/ip_persistence/conftest.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+from libs.net.netattachdef import CNIPluginBridgeConfig, NetConfig, NetworkAttachmentDefinition
+from libs.net.vmspec import lookup_iface_status
+from libs.vm.vm import BaseVirtualMachine
+from tests.network.non_functional.ip_persistence.libippersistence import (
+    LINUX_BRIDGE_IFACE_NAME,
+    linux_bridge_vm,
+)
+from utilities.constants import LINUX_BRIDGE, WORKER_NODE_LABEL_KEY
+
+IP_PERSISTENCE_BRIDGE_DEVICE_NAME = "br1-test"
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/non_functional/ip_persistence/conftest.py]: Done

File: tests/network/non_functional/vmi_interfaces_stability/conftest.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+@pytest.fixture(scope="module")
+def bridge_nad(
+    admin_client: DynamicClient,
+    namespace: Namespace,
+    bridge_nncp: Generator[libnncp.NodeNetworkConfigurationPolicy],
+) -> Generator[NetworkAttachmentDefinition]:
+    config = NetConfig(
+        name="test-bridge-network",
+        plugins=[CNIPluginBridgeConfig(bridge=IP_PERSISTENCE_BRIDGE_DEVICE_NAME)],
```
Comment (azhivovk) [reply tests/network/non_functional/vmi_interfaces_stability/conftest.py]: I'll keep it, thanks

**Comment by openshift-virtualization-qe-bot-3** [comment]
@coderabbitai
<details>
  <summary>Test execution plan request details</summary>

  CRITICAL: You MUST post an inline review comment on the first changed line of the first file.
  The inline comment should contain the Test Execution Plan with the smoke tests recommendation.
  Do NOT submit a formal review - just post the inline comment directly.

  As an expert software testing engineer, analyze all modified files in this PR and create a targeted test execution plan.
  You will post an inline review comment with the test execution plan on the first changed file.
  If you fail to run or post a comment, retry.

  **Analysis Requirements:**

  1. Examine code changes in each modified file
  2. Identify affected code paths, functions, and classes
  3. Analyze pytest-specific elements: fixtures (scope, dependencies), parametrization, markers, conftest changes
  4. Trace test dependencies through imports, shared utilities, and fixture inheritance
  5. Detect new tests introduced in the PR
  6. **Smoke Test Impact Analysis:** Determine if any changes could affect smoke tests by checking:
     - Changes to files/functions used by tests marked with `@pytest.mark.smoke`
     - Changes to fixtures or utilities imported by smoke tests
     - Changes to conftest.py files that may affect smoke test execution
     - Changes to core infrastructure code (utilities/, libs/) that smoke tests depend on

  **Your deliverable:**
  Your change request comment will be based on the following requirements:

  **Test Execution Plan**

  - **Run smoke tests: True / False** _(Based on your smoke test impact analysis - True if any changes may affect smoke tests)_
  - `path/to/test_file.py` - When the entire test file needs verification
  - `path/to/test_file.py::TestClass::test_method` - When specific test(s) needed
  - `path/to/test_file.py::test_function` - When specific test(s) needed
  - `-m marker` - When specific marker(s) can be used to cover multiple cases.

  **Guidelines:**

  - Include only tests directly affected by the changes
  - Use a full file path only if ALL tests in that file require verification
  - Use file path + test name if only specific tests are needed
  - If a test marker can cover multiple files/tests, provide the marker
  - Balance coverage vs over-testing - Keep descriptions minimal

  **Hardware-Related Checks (SR-IOV, GPU, DPDK):**

  When PR modifies fixtures for hardware-specific resources:
  - **Collection Safety**: Fixtures MUST have existence checks (return `None` when hardware unavailable)
  - **Test Plan**: MUST verify both WITH and WITHOUT hardware:
    - Run affected tests on cluster WITH hardware
    - Verify collection succeeds on cluster WITHOUT hardware

  **CRITICAL WORKFLOW COMPLETION RULES:**

  When responding to this test execution plan request, you MUST follow these rules EXACTLY:

  1. **YOUR ONLY DELIVERABLE**: Post an inline review comment containing the test execution plan on the first changed line
  2. **THEN STOP IMMEDIATELY** - Do NOT generate any additional response
  3. **FALLBACK ONLY**: If submitting the review fails after retrying, post as a regular PR comment
  4. **SILENCE = SUCCESS**: After successfully submitting the review, your task is complete. No confirmation needed.

  **ABSOLUTE PROHIBITIONS** (violating these creates empty/meaningless reviews):
  - ❌ Do NOT post acknowledgment messages like "Test execution plan posted", "Review posted successfully", "I've successfully posted"
  - ❌ Do NOT mention review IDs, URLs, or confirmation of posting in the PR thread
  - ❌ Do NOT add any follow-up comments after submitting the review
  - ❌ Do NOT reply to confirm task completion
  - ❌ Do NOT explain what you did - just do it and stop

  **Remember**: The pull request review is visible to users. Additional comments are redundant noise.

  CRITICAL: Before setting "Run smoke tests: True", you MUST use shell scripts to:
    - Trace the actual fixture dependency chain from smoke tests to changed fixtures
    - Verify that smoke tests actually import/use changed utilities or functions
    - Confirm the dependency path exists; do NOT assume based on scope or semantics
    - Be conservative: Session-scoped fixtures or infrastructure-sounding names do NOT
      automatically mean smoke test impact. Only flag smoke test impact when you can
      demonstrate a concrete dependency path.
    WRONG: "This session-scoped storage fixture might affect smoke tests"
    RIGHT: "Smoke test X uses fixture Y, which depends on the changed fixture Z"

</details>

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+    vm: BaseVirtualMachine,
+    ipv4_supported_cluster: bool,
+    ipv6_supported_cluster: bool,
+) -> None:
+    primary_network = lookup_primary_network(vm=vm)
+    secondary_ips = secondary_iface_ips(
+        ipv4_supported_cluster=ipv4_supported_cluster,
+        ipv6_supported_cluster=ipv6_supported_cluster,
+    )
+    spec_interfaces = vm.instance.spec.template.spec.domain.devices.interfaces
```
[the remark this answers is not in the document]
Comment (Anatw) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: If everything is working as expected :) The type system still allows None, and it's better to have a clear failure msg in case of an error, but - we control the creation, so it’s effectively guaranteed... You can resolve this.

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    for iface in spec_interfaces:
+        if iface.name == primary_network.name:
+            lookup_iface_status(vm=vm, iface_name=iface.name)
+        else:
+            lookup_iface_status(
+                vm=vm,
+                iface_name=iface.name,
+                predicate=lambda iface_status: (
+                    "guest-agent" in iface_status["infoSource"]
+                    and iface_status["ipAddress"]
```
Comment (EdDev) [inline tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: This is not needed, you are checking all IPs one line below.

Comment (azhivovk) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: Fixed

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+def secondary_iface_ips(ipv4_supported_cluster: bool, ipv6_supported_cluster: bool) -> list[str]:
+    ips = []
+    if ipv4_supported_cluster:
+        ips.append(random_ipv4_address(net_seed=0, host_address=1))
+    if ipv6_supported_cluster:
+        ips.append(random_ipv6_address(net_seed=0, host_address=1))
+    return ips
+
+
+def wait_for_ifaces_stable(
```
Comment (rnetser) [inline tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: nit: adjective before noun

```suggestion
def wait_for_stable_ifaces
```

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
+import ipaddress
```
[the remark this answers is not in the document]
Comment (azhivovk) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: I also find it a bit unreadable, I can create a follow-up PR to rename all net libs and add underscores

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py (hunk 19 above)
Comment (azhivovk) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: Done

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: I will not approve libs with underscore, unless you show me STD libs that use it as well (common used ones, not some edge cases).

Gemini said
> In the Python Standard Library, underscores in public-facing module names are actually quite rare.
> 
> The Standard Library follows an older naming tradition called "joinedwords" (e.g., datetime, tempfile, multiprocessing) rather than the modern "snake_case" (data_science_module) that you see in third-party libraries.

The main problem I see with packages of libs with their modules is that they looks terrible when used in code. One should assume that the module appears in code for context, e.g. `libfoo.do_foo()`.

Yes, exceptions apply, but this is my proffered format for these ones.

This does not apply for test packages and modules.

**Comment by azhivovk** [comment]
Change:
- [Rename](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930/changes/d59f21a2a40186a846fc3a3e83ec3636b5b9a72b#diff-f7c9b2bf73b7e9c34554fa31cee6f7e8fdd8b0a20f411cfa5fb5619e268559edR99) wait_for_ifaces_stable -> wait_for_stable_ifaces
- [VMI](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930/changes/d59f21a2a40186a846fc3a3e83ec3636b5b9a72b#diff-f7c9b2bf73b7e9c34554fa31cee6f7e8fdd8b0a20f411cfa5fb5619e268559edR124) is VirtualMachineInstance in assert_interfaces_stable

**Comment by EdDev** [comment]
> Change:
> 
> * [Rename](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930/changes/d59f21a2a40186a846fc3a3e83ec3636b5b9a72b#diff-f7c9b2bf73b7e9c34554fa31cee6f7e8fdd8b0a20f411cfa5fb5619e268559edR99) wait_for_ifaces_stable -> wait_for_stable_ifaces
> * [VMI](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930/changes/d59f21a2a40186a846fc3a3e83ec3636b5b9a72b#diff-f7c9b2bf73b7e9c34554fa31cee6f7e8fdd8b0a20f411cfa5fb5619e268559edR124) is VirtualMachineInstance in assert_interfaces_stable

Change: https://github.com/RedHatQE/openshift-virtualization-tests/compare/faed8cce97eed780a9982a1466bf4670de65a5f0..d59f21a2a40186a846fc3a3e83ec3636b5b9a72b

/approve
/lgtm

File: tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: this is why there is no reason to call the module with the full name; it is clear where they belong
Quoing gemini is nice but it all really depends on the prompt you use; if you ask it about standards it will point you to the standard naming conventions which - 1. use underscores 2. follow the sen of python

@azhivovk when introducing new code, the fixes should be in the same pr (and if additional, follow up pr is needed to fix other palces, it is ok). i know there was a habbit for using follow up prs for new code to ease the onborading but this should not be the case anymore

[the remark this answers is not in the document]
Comment (EdDev) [reply tests/network/l2_bridge/vmi_interfaces_stability/libifacestability.py]: I used gemini as a reasoning expressing my opinion.

You can call it `libtest` or `libtools` if you like, or some other general name that can be used all over.
These libs should not be shared between test packages, so it should be ok.
I am ok with it as a middle ground.

**Comment by openshift-virtualization-qe-bot-3** [comment]
@coderabbitai
<details>
  <summary>Test execution plan request details</summary>

  CRITICAL: You MUST post an inline review comment on the first changed line of the first file.
  The inline comment should contain the Test Execution Plan with the smoke tests recommendation.
  Do NOT submit a formal review - just post the inline comment directly.

  As an expert software testing engineer, analyze all modified files in this PR and create a targeted test execution plan.
  You will post an inline review comment with the test execution plan on the first changed file.
  If you fail to run or post a comment, retry.

  **Analysis Requirements:**

  1. Examine code changes in each modified file
  2. Identify affected code paths, functions, and classes
  3. Analyze pytest-specific elements: fixtures (scope, dependencies), parametrization, markers, conftest changes
  4. Trace test dependencies through imports, shared utilities, and fixture inheritance
  5. Detect new tests introduced in the PR
  6. **Smoke Test Impact Analysis:** Determine if any changes could affect smoke tests by checking:
     - Changes to files/functions used by tests marked with `@pytest.mark.smoke`
     - Changes to fixtures or utilities imported by smoke tests
     - Changes to conftest.py files that may affect smoke test execution
     - Changes to core infrastructure code (utilities/, libs/) that smoke tests depend on

  **Your deliverable:**
  Your change request comment will be based on the following requirements:

  **Test Execution Plan**

  - **Run smoke tests: True / False** _(Based on your smoke test impact analysis - True if any changes may affect smoke tests)_
  - `path/to/test_file.py` - When the entire test file needs verification
  - `path/to/test_file.py::TestClass::test_method` - When specific test(s) needed
  - `path/to/test_file.py::test_function` - When specific test(s) needed
  - `-m marker` - When specific marker(s) can be used to cover multiple cases.

  **Guidelines:**

  - Include only tests directly affected by the changes
  - Use a full file path only if ALL tests in that file require verification
  - Use file path + test name if only specific tests are needed
  - If a test marker can cover multiple files/tests, provide the marker
  - Balance coverage vs over-testing - Keep descriptions minimal

  **Hardware-Related Checks (SR-IOV, GPU, DPDK):**

  When PR modifies fixtures for hardware-specific resources:
  - **Collection Safety**: Fixtures MUST have existence checks (return `None` when hardware unavailable)
  - **Test Plan**: MUST verify both WITH and WITHOUT hardware:
    - Run affected tests on cluster WITH hardware
    - Verify collection succeeds on cluster WITHOUT hardware

  **CRITICAL WORKFLOW COMPLETION RULES:**

  When responding to this test execution plan request, you MUST follow these rules EXACTLY:

  1. **YOUR ONLY DELIVERABLE**: Post an inline review comment containing the test execution plan on the first changed line
  2. **THEN STOP IMMEDIATELY** - Do NOT generate any additional response
  3. **FALLBACK ONLY**: If submitting the review fails after retrying, post as a regular PR comment
  4. **SILENCE = SUCCESS**: After successfully submitting the review, your task is complete. No confirmation needed.

  **ABSOLUTE PROHIBITIONS** (violating these creates empty/meaningless reviews):
  - ❌ Do NOT post acknowledgment messages like "Test execution plan posted", "Review posted successfully", "I've successfully posted"
  - ❌ Do NOT mention review IDs, URLs, or confirmation of posting in the PR thread
  - ❌ Do NOT add any follow-up comments after submitting the review
  - ❌ Do NOT reply to confirm task completion
  - ❌ Do NOT explain what you did - just do it and stop

  **Remember**: The pull request review is visible to users. Additional comments are redundant noise.

  CRITICAL: Before setting "Run smoke tests: True", you MUST use shell scripts to:
    - Trace the actual fixture dependency chain from smoke tests to changed fixtures
    - Verify that smoke tests actually import/use changed utilities or functions
    - Confirm the dependency path exists; do NOT assume based on scope or semantics
    - Be conservative: Session-scoped fixtures or infrastructure-sounding names do NOT
      automatically mean smoke test impact. Only flag smoke test impact when you can
      demonstrate a concrete dependency path.
    WRONG: "This session-scoped storage fixture might affect smoke tests"
    RIGHT: "Smoke test X uses fixture Y, which depends on the changed fixture Z"

</details>

**Comment by rnetser** [comment]
> > [Change](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930/changes/cbba97ba7ef14ea835a6d19ebca4aa57c5db2c43#diff-e1ce98ec8220a238e2d59f13a5d3f6a2e5cea4d1bb9c140c4c1ea80a824777e3): Rename new lib to lib_helpers.py as discussed offline
> 
> As discussed in [1] and in many passed discussions from when the `libs` concept has been introduced, I will not accept these names. Said that, I am aware of the root approver opinion and therefore overriding this vote is legit from my side.
> 
> All other changes here looks good to me.
> 
> [1] [#3930 (comment)](https://github.com/RedHatQE/openshift-virtualization-tests/pull/3930#discussion_r2901576581)

following this thread; as discussed, I am approving and merging this PR
